streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignor.java

```diff
@@ -0,0 +1,544 @@
+                        } else {
+                            unassignedStandbyTasks.put(task, remainingStandbys - 1);
+                        }
+                    }
+                }
+            }
+        }
+        return unassignedActiveTasks.isEmpty() && unassignedStandbyTasks.isEmpty();
+    }
+
+    /**
+     * @return true if this client is caught-up for this task, or the task has no caught-up clients
+     */
+    boolean taskIsCaughtUpOnClient(final TaskId task, final ID client) {
+        boolean hasNoCaughtUpClients = true;
+        final SortedSet<RankedClient<ID>> rankedClients = statefulTasksToRankedCandidates.get(task);
+        if (rankedClients == null) {
+            return true;
+        }
+        for (final RankedClient rankedClient : rankedClients) {
+            if (rankedClient.rank() <= 0L) {
+                if (rankedClient.clientId().equals(client)) {
+                    return true;
+                } else {
+                    hasNoCaughtUpClients = false;
+                }
+            }
+
+            // If we haven't found our client yet, it must not be caught-up
+            if (rankedClient.rank() > 0L) {
+                break;
+            }
+        }
+        return hasNoCaughtUpClients;
+    }
+
+    /**
+     * Rankings are computed as follows, with lower being more caught up:
+     *      Rank -1: active running task
+     *      Rank 0: standby or restoring task whose overall lag is within the acceptableRecoveryLag bounds
+     *      Rank 1: tasks whose lag is unknown, eg because it was not encoded in an older version subscription
+     *      Rank 1+: all other tasks are ranked according to their actual total lag
+     * @return Sorted set of all client candidates for each stateful task, ranked by their overall lag. Tasks are
+     */
+    static <ID extends Comparable<ID>> SortedMap<TaskId, SortedSet<RankedClient<ID>>> buildClientRankingsByTask(final Set<TaskId> statefulTasks,
+                                                                                                                final Map<ID, ClientState> clientStates,
+                                                                                                                final long acceptableRecoveryLag) {
+        final SortedMap<TaskId, SortedSet<RankedClient<ID>>> statefulTasksToRankedCandidates = new TreeMap<>();
+
+        for (final TaskId task : statefulTasks) {
+            final SortedSet<RankedClient<ID>> rankedClientCandidates = new TreeSet<>();
+            statefulTasksToRankedCandidates.put(task, rankedClientCandidates);
+
+            for (final Map.Entry<ID, ClientState> clientEntry : clientStates.entrySet()) {
+                final ID clientId = clientEntry.getKey();
+                final long taskLag = clientEntry.getValue().lagFor(task);
+                final long clientRank;
+                if (taskLag == Task.LATEST_OFFSET) {
+                    clientRank = Task.LATEST_OFFSET;
```

Comment:
I think we should still prioritize the client that had the active task over one with a caught-up standby. For one thing, with KIP-429 we have to revoke an active task before moving it to a new client in a followup rebalance. This means deadtime for that active task between being closed on one client, waiting for another rebalance, and finally being recreated from a standby on a new client. We also lose the cache, buffers, etc

Reply:
That's fair. My concern about the impact was whether it results in non-termination of the probing rebalance cycle, if we always prefer to re-assign the prior active and always propose to move the task to the same caught-up standby, but never consider just giving the active to the caught-up standby, since there is a prior active.

Reply:
Actually I'm pretty concerned about the same thing, but I'm not sure I see how ranking previous active tasks saves us from getting trapped in the rebalancing cycle. Might be worth chatting about offline..

Reply:
I suppose that this, among other things, will become clear when we add some integration and system tests.